controller/job/jobcontroller.go

```diff
@@ -0,0 +1,179 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import (
+	"context"
+	"sync"
+
+	"github.com/crunchydata/postgres-operator/config"
+	"github.com/crunchydata/postgres-operator/ns"
+	"github.com/crunchydata/postgres-operator/operator"
+	log "github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/batch/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Controller holds the connections for the controller
+type Controller struct {
+	JobConfig          *rest.Config
+	JobClient          *rest.RESTClient
+	JobClientset       *kubernetes.Clientset
+	Ctx                context.Context
+	informerNsMutex    sync.Mutex
+	InformerNamespaces map[string]struct{}
+}
+
+const (
+	patchResource = "pgtasks"
+	patchURL      = "/spec/status"
+)
+
+// Run starts an pod resource controller
+func (c *Controller) Run() error {
+
+	err := c.watchJobs(c.Ctx)
+	if err != nil {
+		log.Errorf("Failed to register watch for job resource: %v\n", err)
+		return err
+	}
+
+	<-c.Ctx.Done()
+	return c.Ctx.Err()
+}
+
+// watchJobs is the event loop for job resources
+func (c *Controller) watchJobs(ctx context.Context) error {
+	nsList := ns.GetNamespaces(c.JobClientset, operator.InstallationName)
+	log.Debugf("jobController watching %v namespaces", nsList)
+
+	for i := 0; i < len(nsList); i++ {
+		log.Infof("starting job controller for ns [%s]", nsList[i])
+		c.SetupWatch(nsList[i])
+
+	}
+	return nil
+}
+
+// onAdd is called when a postgresql operator job is created and an associated add event is
+// generated
+func (c *Controller) onAdd(obj interface{}) {
+
+	job := obj.(*apiv1.Job)
+
+	//only process jobs with with vendor=crunchydata label
+	labels := job.GetObjectMeta().GetLabels()
+	if labels[config.LABEL_VENDOR] == "crunchydata" {
+		log.Debugf("Job Controller: onAdd ns=%s jobName=%s", job.ObjectMeta.Namespace, job.ObjectMeta.SelfLink)
+		return
+	}
+}
+
+// onUpdate is called when a postgresql operator job is created and an associated update event is
+// generated
+func (c *Controller) onUpdate(oldObj, newObj interface{}) {
+
+	var err error
+	job := newObj.(*apiv1.Job)
+	labels := job.GetObjectMeta().GetLabels()
+
+	//only process jobs with with vendor=crunchydata label
+	if labels[config.LABEL_VENDOR] == "crunchydata" {
+		log.Debugf("[Job Controller] onUpdate ns=%s %s active=%d succeeded=%d conditions=[%v]",
+			job.ObjectMeta.Namespace, job.ObjectMeta.SelfLink, job.Status.Active, job.Status.Succeeded,
+			job.Status.Conditions)
+	}
```

Comment:
Should this be `!= crunchydata → early return` ?

Reply:
Updated to return early if the crunchy vendor label isn't found.